Allow Artists to show pixel data in cursor display

blink1073 opened the pull request:
Also adds implementation for `AxesImage`.  Closes #3984.
Inspired by [mpldatacursor](https://github.com/joferkington/mpldatacursor).

![screenshot from 2015-02-21 12 13 27](https://cloud.githubusercontent.com/assets/2096628/6315292/352706d0-b9c3-11e4-9d39-89c5bfb63ae7.png)

![screenshot from 2015-02-21 12 14 20](https://cloud.githubusercontent.com/assets/2096628/6315294/3b431d1a-b9c3-11e4-9906-e10bd8ffc7c6.png)

tacaswell commented on lib/matplotlib/backend_bases.py:
```diff
@@ -2791,6 +2791,12 @@ def mouse_move(self, event):
                 pass
             else:
+                ax = event.inaxes
+                artists = ax.artists + ax.images + ax.lines
```
Should also look at `ax.collections` (that is where things like scatter live) and maybe `ax.patches`

tacaswell replied:
and the order should be `ax.collections + ax.patches + ax.lines + ax.artists + ax.images` to match the order used for sorting out the rendering order  in `Axes.draw` (around https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/axes/_base.py#L1993 images are special cased around line 2030).

blink1073 replied:
They then get ordered by zorder anyway: https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/axes/_base.py#L2036

tacaswell replied:
Right, but the ordering is a stable sort so ties are broken by which was in the list first.  This is the source of at least one un-fixable (with out major overhaul of the internal data structure, see #3944)  bug ( #409 ).

blink1073 replied:
Ok then. So how to handle composite images?  It does not seem feasible to recreate the composite every time.

blink1073 replied:
The `AxesImage` could composite the specific pixel on demand.

tacaswell replied:
I think to first order, just ignore the compositing and use the top image. 

The in-place compositing is only for some of the vector backends which can't do it them selves (ex pdf) to try and save space (so you don't have multiple overlapping images) but it leads to other bugs as all images get smashed to the same zorder relative to the other lines etc.   In Agg (which backs all but 3 (gtkcairo, macsox and windows) of the interactive backends) the whole `if _do_composite` block is skipped and the compsiting is taken care of at the c++ layer.

blink1073 commented:
How's that?  I noticed `onRemove` was doing something similar and found `figure.hitlist()`, and standardized between the two.

blink1073 commented:
However, it looks like `Axes.get_children` needs to return the same order as `Axes.draw`.  New PR for that?

tacaswell commented:
`onRemove` looks like another candidate (along with `onHilite`) for removal from `FigureCanvasBase` (now that you have simplified it)....

Probably worth having a top-level module for useful event handlers, which I think we greatly reduce the threshold to get people using the event system.  @WeatherGod @joferkington thoughts?

Yes to separate PR for re-ordering `Axes.get_children` as that is a (subtle) api change.

blink1073 commented:
I think we do need to handle compositing.  For instance, if I run this [example](https://github.com/scikit-image/scikit-image/blob/master/viewer_examples/plugins/canny.py), I get the Zdata for the overlay layer.  How about just checking for `alpha=0` and then using the underlying in that case?

blink1073 commented:
Well, the more I think about it the more I agree that the top image should get priority.  The canny value really is the point of the example.

blink1073 commented:
Added `Artist.format_zdata` to centralize formatting.

blink1073 commented:
Do I still "need revision"?

tacaswell commented:
Looks good to me.  I am a tad wary of the `format_z` method which does not use `self`, but that seems like a reasonable place to put it.

I would like @WeatherGod to weigh in.

blink1073 commented:
What do you mean "does not use `self`"?

blink1073 commented:
Meaning it should be a static method?

blink1073 commented:
Either way I think it belongs in the class so it can be overridden.

tacaswell commented:
Yes, it is an instance method which does not refer to it's instance and I can't think of a use case where you would want to (but I may just have poor imagination).  I thought about suggesting to move it to `cbook`, but it is much too specific for that.

blink1073 commented:
I'd worry that adding a `@staticmethod` would give people the impression they could not subclass it as an instance method (I checked, and you can).

tacaswell commented:
This also needs an entry in https://github.com/matplotlib/matplotlib/tree/master/doc/users/whats_new

blink1073 commented:
Done.

tacaswell commented:
@matplotlib/developers Can this get eyes from someone else but me?

This addresses a long standing and frequent feature request in a pleasingly general way.

efiring commented on lib/matplotlib/artist.py:
```diff
@@ -903,6 +903,27 @@ def matchfunc(x):
+        """
+        Get the zdata for a given event, as a string message
+        """
+        return ''
```
Why do you want to use a string for this?  Shouldn't "data" always be numeric?

blink1073 replied:
This allows the `Artist` to control the display of numbers, similar to the existing `ax.format_coord`.

joferkington replied:
I agree with @efiring on this one.  I can think of a lot of use cases (e.g. interactive sampling) where it would be very convenient to access a "raw" value.  Returning a numerical value would make this method useful for a lot more than just display.

You already have a separate method for the formatting, so it's only slightly more verbose to flip the function calls on their head and make the displayed string using `artist.format_zdata(artist.get_zdata())`.  In my opinion, this would be a better approach overall.

blink1073 replied:
Ha!  I originally meant to do that...

tacaswell commented:
@joferkington again.  Sorry to pester, but I see you are doing OSS stuff today and would like your feed back on this :smile:

joferkington commented:
@tacaswell - Absolutely!  Sorry I've been so out-of-touch with things lately, as well.

pelson commented on lib/matplotlib/image.py:
```diff
@@ -687,6 +687,21 @@ def get_extent(self):
+        arr = self.get_array()
+        data_extent = mtransforms.Bbox([[ymin, xmin], [ymax, xmax]])
+        array_extent = mtransforms.Bbox([[0, 0], arr.shape[:2]])
+        trans = (mtransforms.BboxTransformFrom(data_extent) +
```
FWIW There is a BboxTransform which gives you Bbox -> another Bbox without having to go to a unit Bbox.

pelson commented on lib/matplotlib/backend_bases.py:
```diff
@@ -2791,6 +2785,12 @@ def mouse_move(self, event):
+                    artists.remove(event.inaxes.patch)
+                artists.sort(key=lambda x: x.zorder)
+                if artists:
+                    s += artists[-1].get_zdata(event)
```
Why only the last one? Is it a thing to want to have more than one zdata at a time?

joferkington replied:
@pelson  - By default, there isn't enough room in the cursor window for more than one.

If nothing else, Tkinter does odd things when the string displayed in the cursor becomes larger than the box it's in.  (The window resizes slightly, and it gives a visually jarring "jump" effect.)

pelson commented on lib/matplotlib/image.py:
```diff
@@ -687,6 +687,21 @@ def get_extent(self):
+        y, x = event.ydata, event.xdata
+        i, j = trans.transform_point([y, x]).astype(int)
+        z = arr[i, j]
+        return 'z=%s' % self.format_zdata(z)
```
Should it be the image's job to say "z=" - and if so, should it say ``pixel-value=``?

pelson commented:
Nice and simple! It would be good to get a few tests in place, but generally happy with the concept.

joferkington commented:
Definitely +1 from me.  

I do think returning a numercial value from `artist.get_zdata` would make the method useful for other purposes.

A test or two would certainly help.  You can fire a "fake" event even with the Agg backend and verify that the correct value is extracted.  You'd want to do this with different `origin` and `extent` kwargs, but something vaguely like the following:

```
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.backend_bases import MouseEvent

fig, ax = plt.subplots()
im = ax.imshow(np.arange(100).reshape(10, 10))

x, y = 0, 0
xdisp, ydisp = ax.transData.transform_point([x, y])

event = MouseEvent('motion_notify_event', fig.canvas, xdisp, ydisp)
assert im.get_zdata(event) == 0
```

Either way, nice addition!

blink1073 commented:
I addressed all comments but @pelson's:

"FWIW There is a BboxTransform which gives you Bbox -> another Bbox without having to go to a unit Bbox."

-- I have no idea what you mean.

blink1073 commented:
A thought: we could display as `x=1 y=1 [100]` to save space.

pelson commented:
> -- I have no idea what you mean.

Lol. I do that sometimes :smile: 

```
trans = (mtransforms.BboxTransformFrom(data_extent) +
                 mtransforms.BboxTransformTo(array_extent))
```

Is the same as

```
trans = mtransforms. BboxTransform(boxin=data_extent, boxout=array_extent)
```

blink1073 commented:
Thanks @pelson, I'm a little slow.  :smile:  
I updated the description with some screen shots.

cimarronm commented on lib/matplotlib/artist.py:
```diff
@@ -903,6 +903,27 @@ def matchfunc(x):
+            is_int = isinstance(data, int)
+            data = [data]
+        if is_int:
+            data = ', '.join(item for item in data)
```
How is this working? I don't think `is_int` is ever being set as this line should not work - `join`should give a TypeError if it gets a sequence of `int` and not a sequence of string. Try ```data=[1, 2, 3]; data = ', '.join(item for item in data)``` (BTW, you could have just used `data` instead of `item for item in data`).

I suspect what is going on in your example is that a numpy array is being given and `is_int` is being set to `False` and the else condition is always used. That being the case, then you might as well not even test for int and just use the `{:0.3g}` formatter in all cases.

blink1073 replied:
Agreed, simplified.

cimarronm commented:
A very nice feature to have for images. This is good work @blink1073!

I'm not sure on the inclusion of the `format_pixel_data` in the `artist` class. It really is being used as a static method and not inherently related to the artist class in any way. I think I would favor seeing this included somewhere else that makes better sense or done away with entirely (just incorporate into `backend_bases`).

blink1073 commented:
@cimarronm, I put `format_pixel_data` on the artist so that artists could decide the best way to represent themselves.

tacaswell commented:
I have clearly not thought this out very carefully yet, but I think it would be reasonable to use this to have scatter plots report their size/array values into the mesage.

It might also be cute (but breaks the current naming) to have Line2D objects report their label in the message bar.

This is one of those 'batteries included' things that I think will have a major impact in usability for relatively little work on the dev side.

blink1073 commented:
@tacaswell, would the name be `cursor_data` be a good fit?

tacaswell commented:
I like `cursor_data`.  

`message_data` (`msg_data`?) also seems reasonable given the other internal naming schemes, but is a bit more ambiguous.

dopplershift commented on lib/matplotlib/backend_bases.py:
```diff
@@ -2791,6 +2785,15 @@ def mouse_move(self, event):
+                if event.inaxes.patch in artists:
+                    artists.remove(event.inaxes.patch)
+                artists.sort(key=lambda x: x.zorder)
+                if artists:
```
Something seems odd about checking if `artists` is empty *after* sorting the list. I know (provided `artists` is a list) it won't error out, but still seems to me that the sort should go ahead and move within the `if` block.

Also, +1 for removing ye olden style DSU in favor of `sort(key=)`.

blink1073 replied:
Good call, done.

dopplershift commented:
+1 on the name of `cursor_data`. This is a nice feature (finally) to have and a very simple implementation.
